### PR TITLE
Select2 products list in voucher form 

### LINKED ISSUE #1398: Use select2 when applying voucher to only one product
It would be nice if select input in vouchers menu used `select2` to pick product name - at this moment all products are dumped to input, which causes it to be inconvenient if product list is longer than few items.

### PR DESCRIPTION
**Pull request by mad-anne**
Replaces default ChoiceField for project's custom AjaxSelect2ChoiceField in voucher form.

Closes #1398

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] Python code quality checks pass: `pycodestyle`, `pydocstyle`, `pylint`.
1. [x] JavaScript code quality checks pass: `eslint`.

### CODE REVIEW

File: saleor/dashboard/product/views.py [hunk 1]
```diff
@@ -711,3 +711,21 @@ def get_variant_label(variant):
         for variant in queryset
     ]
     return JsonResponse({'results': variants})
+
+
+@staff_member_required
```
Comment (akjanik) [inline saleor/dashboard/product/views.py]: Shouldn't we guard those views with permissions similar to non-ajax views?

Comment (elwoodxblues) [reply saleor/dashboard/product/views.py]: Good catch. I'm not sure what restrictions are set on the "add voucher" view, but if there is the ability to view the products, then the person having rights to edit vouchers also needs rights to view the products.

**Comment by elwoodxblues** [comment]
@mad-anne I just checked that we have a select with products in the sales form too, so Select2 widget would be welcomed there as well. Also, the list of countries in "Price per country" form may utilize this widget as it is quite long (although it doesn't the need dynamic AJAX prefetching since it's static).
As for the permissions - both voucher and sales views have their own permissions, but their using either products or categories inside. Shouldn't we also require "product" permissions to access those views (cc @koradon @akjanik)?

**Comment by patrys** [comment]
@elwoodxblues I feel that for the country selection in shipping pricing we should open a separate issue as the country choices there should allow multiple selections.

**Comment by akjanik** [comment]
@elwoodxblues Depends what we want to achieve. Ideally permissions should be as intuitive as possible; I mean, we don't want someone to be stuck at point where he can't access view, but he/she has "proper" permission, because someone forgot give additional one.

**Comment by elwoodxblues** [comment]
@akjanik I agree with that. But in this case, it looks like the vouchers view simply requires two types of permissions. If we require only "vouchers" permission, we would leak the list of products. Maybe some kind of dependency between permissions is needed here. E.g. when giving access to the vouchers view, we would also have to check if access to the products list was also given and raise a ValidationError otherwise. @patrys Do you have any thoughts what we could do about this?

**Comment by patrys** [comment]
I think having either of the qualifying permissions (this means either "can view products" or "can manage vouchers") should be enough. Otherwise it's a useless view that appears to be broken (haven't tested this but I assume the AJAX response will return a 403 instead of matching products).

**Comment by elwoodxblues** [comment]
I'm +1 on limiting the list of products only to the public ones when the user has no "can view products" permission.

**Comment by mad-anne** [comment]
List of products is already filtered in this view by available ones, so all returned products are published in the storefront. We can conditionally show all products if the user has ```view_product``` permission, but it will enable creating vouchers and sales also to currently unavailable products. If it would be right in your opinion (I suppose it makes sense at all), i will change it.

**Comment by elwoodxblues** [comment]
@mad-anne Adding vouchers or sales for unavailable products shouldn't have any effect. Since those products can't be added to cart, there won't be any way to use the voucher or sale in the storefront.

**Comment by mad-anne** [comment]
@elwoodxblues but during sale some of products may become available, so staff users should have a possiblity to add them to a sale even if they are not visible in the storefront yet. I changed code to fit this rule.

**Comment by patrys** [comment]
I agree, it should be possible to prepare the necessary vouchers ahead of publishing products.

File: saleor/core/forms.py [hunk 2]
```diff
@@ -7,11 +7,15 @@ class AjaxSelect2ChoiceField(forms.ChoiceField):
     fetch_data_url - specifies url, from which select2 will fetch data
     """
 
-    def __init__(self, fetch_data_url='', initial=None, *args, **kwargs):
+    def __init__(
+            self, fetch_data_url='', initial=None, many=False, *args,
```
[the remark this answers is not in the document]
Comment (mad-anne) [reply saleor/core/forms.py]: I took a naming patten from DRF, but maybe it is not enough meaningful in Django. I think second solution with inheriting field is better, I'll change it.

**Pull request merged**